File: frontend/IOS_Swift/Glass/DuplicateTextSuppressor.swift
import Foundation

final class DuplicateTextSuppressor {
    private var lastNormalizedText = ""
    private var lastSpokenDate: Date = .distantPast
    private let cooldown: TimeInterval = 8.0

    func shouldSpeak(_ text: String, now: Date = Date()) -> Bool {
        let normalized = normalize(text)
        guard !normalized.isEmpty else { return false }

        if isDuplicate(normalized, lastNormalizedText) {
            return false
        }

        if now.timeIntervalSince(lastSpokenDate) < cooldown {
            let previousWords = Set(lastNormalizedText.split(separator: " "))
            let newWords = Set(normalized.split(separator: " "))
            let overlap = previousWords.intersection(newWords).count
            if overlap > 0 {
                return false
            }
        }

        lastNormalizedText = normalized
        lastSpokenDate = now
        return true
    }

    private func normalize(_ text: String) -> String {
        text
            .lowercased()
            .components(separatedBy: CharacterSet.alphanumerics.inverted)
            .filter { !$0.isEmpty }
            .joined(separator: " ")
    }

    private func isDuplicate(_ first: String, _ second: String) -> Bool {
        guard !first.isEmpty, !second.isEmpty else { return false }
        if first == second { return true }
        if first.contains(second) || second.contains(first) { return true }

        let firstWords = Set(first.split(separator: " "))
        let secondWords = Set(second.split(separator: " "))
        guard !firstWords.isEmpty, !secondWords.isEmpty else { return false }

        let overlap = firstWords.intersection(secondWords).count
        let union = firstWords.union(secondWords).count
        return Double(overlap) / Double(union) >= 0.84
    }
}

File: frontend/Android/app/src/main/java/com/kojingapla/glass/DuplicateTextSuppressor.kt
package com.kojingapla.glass

class DuplicateTextSuppressor {
    private var lastNormalizedText = ""
    private var lastSpokenMs = 0L
    private val cooldownMs = 8_000L

    fun shouldSpeak(text: String, nowMs: Long = System.currentTimeMillis()): Boolean {
        val normalized = normalize(text)
        if (normalized.isEmpty()) return false

        if (isDuplicate(normalized, lastNormalizedText)) return false

        if (nowMs - lastSpokenMs < cooldownMs) {
            val previousWords = lastNormalizedText.split(" ").filter { it.isNotEmpty() }.toSet()
            val newWords = normalized.split(" ").filter { it.isNotEmpty() }.toSet()
            if (previousWords.intersect(newWords).isNotEmpty()) return false
        }

        lastNormalizedText = normalized
        lastSpokenMs = nowMs
        return true
    }

    private fun normalize(text: String): String =
        text.lowercase()
            .replace(Regex("[^\\p{IsAlphabetic}\\p{IsDigit}]+"), " ")
            .trim()
            .replace(Regex("\\s+"), " ")

    private fun isDuplicate(first: String, second: String): Boolean {
        if (first.isEmpty() || second.isEmpty()) return false
        if (first == second || first.contains(second) || second.contains(first)) return true

        val firstWords = first.split(" ").toSet()
        val secondWords = second.split(" ").toSet()
        val overlap = firstWords.intersect(secondWords).size
        val union = firstWords.union(secondWords).size
        return union > 0 && overlap.toDouble() / union.toDouble() >= 0.84
    }
}
